### PR TITLE
Take and Take Grad

### PR DESCRIPTION
**Pull request by antinucleon**

### CODE REVIEW

**Comment by antinucleon** [comment]
@tqchen

**Comment by piiswrong** [comment]
looks good to me

File: mshadow/extension/take.h [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  return Plan<TakeExp<IndexExp, SrcExp, DType>, DType>(exp);
+}
+
+template<int dim, typename IndexExp, typename SrcExp, typename DType>
+struct ShapeCheck<dim, TakeExp<IndexExp, SrcExp, DType> > {
+  inline static Shape<dim>
+  Check(const TakeExp<IndexExp, SrcExp, DType> &t) {
+    CHECK(dim == 2)
+      << "TakeExp only support 2D output";
+    Shape<1> dshape = ShapeCheck<1, IndexExp>::Check(t.index_);
```
Comment (tqchen) [inline mshadow/extension/take.h:80]: Maybe need a more general form, think about the case where input is batched

File: mshadow/extension/take.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+ public:
+  explicit Plan(const TakeExp<IndexExp, SrcExp, DType> &e)
+    : index_(MakePlan(e.index_)), src_(MakePlan(e.src_)) {
+  }
+
+  // TODO(xx): discuss W shape: in * out or out * in
+  // Now I use in * out
+  MSHADOW_XINLINE DType Eval(index_t y, index_t x) const {
+    index_t idx = static_cast<index_t>(index_.Eval(0, y));
+    return static_cast<DType>(src_.Eval(idx, x));
```
Comment (tqchen) [inline mshadow/extension/take.h:60]: LGTM

**Pull request merged**